rfcs/007 - Alternative Ways to Create Mappers.md

```diff
@@ -0,0 +1,122 @@
+# 007 - Alternative Ways to Create Mappers
+
+## Current Status
+
+### Proposed
+
+2023-02-07
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Full Name <email@example.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Mohammad Teimori Pabandi <m@iasql.com>
+
+## Summary
+
+Having Pure-SQL modules would help us expand more. One key advantage is that we don't need a PR on our engine to add coverage for a new service, we just use a syntax like this and everything JustWorks™.
+```postgresql
+SELECT iasql_install('https://github.com/someModuleDeveloper/awsEc2Module');
+```
+That's where we started, but it might be useful to assess different paths we might take for creation of an IaSQL module/mapper.
+
+## Proposal
+
+To create a module, we need a mapper. I'll continue this RFC with my proposed solution for mappers, since a module is a set of mappers, and entities (with a bit of oversimplification).
+
+First, let's see what we need to create a mapper:
+- A cloud id function
+- An equality checker function
+- An entity, that the mapper manages it
+- Cloud CRUD functions
+- Ability to call other mappers' CRUD functions
```

Comment:
We should quantify how many of these are actually necessary once we're inside of Postgres where we can literally just `SELECT * FROM other_entity WHERE some_condition = TRUE`. And perhaps triggering Cloud CRUD functions ought to be done by INSERT/UPDATE/DELETE on the database and letting the apply/commit logic just deal with it?

Reply:
> And perhaps triggering Cloud CRUD functions ought to be done by INSERT/UPDATE/DELETE on the database and letting the apply/commit logic just deal with it?

I think we still need to be able to call the mappers' functions directly. Yes, we can trigger cloud CRUD functions by manipulating the data inside Postgres, but we can't wait for the result (or I don't know of a way to do so) of those functions. That's kind of event-driven, but without the callback yet.

If a module's cloud CRUD functions depend on these functions on another module, then we'll need to provide that interface.